### PR TITLE
Make Response::to_json preserve io::Error of ErrorKind::TimedOut

### PR DESCRIPTION
**Pull request by algesten**

### CODE REVIEW

**Comment by algesten** [comment]
@jsha Does this look reasonable to you? serde_json doesn't let us get the actual underlying struct, so to "preserve" the io::Error, I clone to be as good as possible.

I'll add a test case if you think it looks ok.

File: src/request.rs [hunk 1]
```diff
@@ -613,7 +613,10 @@ impl Request {
     ///     .set_tls_connector(tls_connector.clone());
     /// ```
     #[cfg(all(feature = "native-tls", not(feature = "tls")))]
-    pub fn set_tls_connector(&mut self, tls_connector: Arc<native_tls::TlsConnector>) -> &mut Request {
+    pub fn set_tls_connector(
```
Comment (jsha) [inline src/request.rs:616]: This looks like an unrelated formatting change. Usually I try to put formatting changes in their own commit so they're easier to understand in a `git blame`.

**Comment by algesten** [comment]
@jsha Making the test case I realised we needed to cover the `WouldBlock` scenario too. I added some more commits around that. Notably I ended up blanket cloning all `io::Error`. Looks ok?

**Comment by algesten** [comment]
Ok! I'll make another PR to land first for that.

**Comment by algesten** [comment]
@jsha I believe this one is ready to go in now (take 2).
